sandberg_christy_ndvi.py

```diff
@@ -0,0 +1,700 @@
+# 
+# Create a plot of the mean normalized difference vegetation index (NDVI) for the two different sites in the United States across the year: 
+# 
+# * NDVI on the x axis and formatted dates on the y for both NEON sites on one figure/axis object.
+# * Each site should be identified with a different color in the plot and legend.
+# * The final plot **data should be cleaned to remove the influence of clouds**.
+# * Be sure to include appropriate title and axes labels.
+# 
+# Add additional cells as needed for processing data (e.g. defining functions, etc), but be sure to:
+# * follow the instructions in the code cells that have been provided to ensure that you are able to use the sanity check tests that are provided. 
+# * include only the plot code in the cell identified for the final plot code below
+
+# ## Task 1: 
+# 
+# In the cell below, create a single dataframe containing MEAN NDVI, the site name, 
+# and the date of the data for the HARV site 
+# scene `HARV/landsat-crop/LC080130302017031701T1-SC20181023151837`.  The column names for the  final
+# DataFrame should be`mean_ndvi`, and `site`, and the data should be **indexed on the date**. 
+# 
+# Use the functions that we reviewed in class (or create your own versions of them) to implement your code
+# 
+# ### In the Cell below Place  All Functions Needed to Run this Notebook (20 points)
+
+# The two functions I've created for this assignment are:
+# 
+# 1. **'process_single_scene'** - this function will collect four variables needed 
+# from each scene to populate the final dataframe with 'site', 'date' and 
+# 'mean_ndvi'. 
+# 
+#     1. site name
+#     2. data acquisition date
+#     3. cloud mask layer from scene's 'qa' file
+#     4. list of red and near-infrared bands
+#     
+#     This function will open the red and near-infrared bands needed for the NDVI 
+#     calculation, clipping to a crop shape that is identified earlier in the 
+#     function. 
+#     It will also remove any values outside of a given valid range of values (if 
+#     that valid range is provided to the function). The actual NDVI calculation 
+#     occurs outside the 'process_single_scene' function.
+#     
+# 
+# 2. **'calculate_masked_ndvi'** - this function will calculate the NDVI from two 
+# bands provided in the 'ndvi_list', and then mask the NDVI values from a cloud 
+# mask layer generated from the 'pixel_qa' variable provided by the 
+# 'process_single_scene' function and a 'cloud_vals' list that can be maintained 
+# inside the function.
+
+# In[4]:
+
+
+### DO NOT REMOVE THIS LINE OR EDIT / MOVE THIS CELL ###
+start_time = datetime.now()
+
+
+# In[5]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+
+# YOUR CODE HERE
+ 
+# function to process single landsat scene in directory 
+def process_single_scene (path_to_scene_dir, valid_range):
```

Comment:
@csandberg303 I think this function is really clever in that you manage to do all the heavy lifting of the directory parsing and the variables that are needed to feed into the next function. The comment/criticism I would like to throw out there is that there are some really targeted/specific lines of code (e.g., 289-295) with respect to the directory structure(s). I wonder if there is a way to write the function such that the specific variables (e.g., {10:18] are actually inputs in the function instead of hard coded. FWIW, I couldn't figure out how to do this on my own, but I wish I knew how!

Reply:
@esgeo I agree that the function relies heavily on consistent file naming and directory structure.  I understand Landsat files follow a naming convention so that the date should always show in the same spot (10:18), but I don't think I can expect to find a directory named 'vector' outside of this current project.